### PR TITLE
Handle checkpoint dirpath suffix in NeptuneLogger

### PR DESCRIPTION
**Pull request by AleksanderWWW**
## What does this PR do?

Fixes #\<issue_number>

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

----
:books: Documentation preview :books:: https://pytorch-lightning--18863.org.readthedocs.build/en/18863/

### CODE REVIEW

File: src/lightning/pytorch/loggers/neptune.py [hunk 1]
```diff
@@ -43,6 +43,12 @@
 _INTEGRATION_VERSION_KEY = "source_code/integrations/pytorch-lightning"
 
 
+def _get_expected_model_path(dir_path: str) -> str:
+    expected_model_path = dir_path
+    expected_model_path = expected_model_path.rstrip(os.path.sep) if expected_model_path else expected_model_path
```
Comment (SiddhantSadangi) [inline src/lightning/pytorch/loggers/neptune.py:48]:
```suggestion
    expected_model_path = dir_path.rstrip(os.path.sep) if dir_path else dir_path
```

Comment (SiddhantSadangi) [inline src/lightning/pytorch/loggers/neptune.py:48]:
```suggestion
    if dir_path and dir_path[-1] in ["/","\\"]:
        expected_model_path = dir_path[:-1]
    else:
        expected_model_path = dir_path
```

**Review by SiddhantSadangi — CHANGES_REQUESTED** [review]

File: src/lightning/pytorch/loggers/neptune.py [hunk 2]
```diff
@@ -43,6 +43,12 @@
...
+    return expected_model_path + os.path.sep
```
Comment (SiddhantSadangi) [inline src/lightning/pytorch/loggers/neptune.py]:
```suggestion
def _get_expected_model_path(dir_path: str) -> str:
    expected_model_path = dir_path

    while expected_model_path and expected_model_path[-1] in ("/", "\\"):
        expected_model_path = expected_model_path[:-1]

    return f"{expected_model_path}{os.path.sep}"
```

File: tests/tests_pytorch/loggers/test_neptune.py [hunk 3]
```diff
@@ -303,3 +304,13 @@ def test_get_full_model_names_from_exp_structure():
     }
     expected_keys = {"lvl1_1/lvl2/lvl3_1", "lvl1_1/lvl2/lvl3_2", "lvl1_2"}
     assert NeptuneLogger._get_full_model_names_from_exp_structure(input_dict, "foo/bar") == expected_keys
+
+
+def test_get_expected_model_path():
+    assert _get_expected_model_path("my_model/checkpoints") == "my_model/checkpoints/"
+    assert _get_expected_model_path("my_model/checkpoints/") == "my_model/checkpoints/"
+
```
Comment (SiddhantSadangi) [inline tests/tests_pytorch/loggers/test_neptune.py:318]:
```suggestion
    assert _get_expected_model_path("my_model/checkpoints") == "my_model/checkpoints/"
    assert _get_expected_model_path("my_model/checkpoints/") == "my_model/checkpoints/"
    assert _get_expected_model_path("my_model/checkpoints//") == "my_model/checkpoints/"
```

File: tests/tests_pytorch/loggers/test_neptune.py [hunk 4]
```diff
@@ -303,3 +304,13 @@ def test_get_full_model_names_from_exp_structure():
...
+
+@patch("os.path.sep", new="\\")
+def test_get_expected_model_path_windows():
+    assert _get_expected_model_path("my_model\\checkpoints\\") == "my_model\\checkpoints\\"
```
Comment (SiddhantSadangi) [inline tests/tests_pytorch/loggers/test_neptune.py:322]:
```suggestion
    assert _get_expected_model_path("my_model\\checkpoints\\") == "my_model\\checkpoints\\"
    assert _get_expected_model_path("my_model\\checkpoints") == "my_model\\checkpoints\\"
```

File: src/lightning/pytorch/loggers/neptune.py [hunk 5]
```diff
@@ -43,6 +43,16 @@
...
 
 
+def _get_expected_model_path(dir_path: str) -> str:
+    expected_model_path = dir_path
+
+    while expected_model_path and expected_model_path[-1] in ("/", "\\"):
+        expected_model_path = expected_model_path[:-1]
+
+    expected_model_path = f"{expected_model_path}{os.path.sep}"
+    return expected_model_path
```
Comment (SiddhantSadangi) [inline src/lightning/pytorch/loggers/neptune.py:53]:
```suggestion
    return f"{expected_model_path}{os.path.sep}"
```

**Review by SiddhantSadangi — CHANGES_REQUESTED** [review]

**Review by SiddhantSadangi — APPROVED** [review]

**Review by SiddhantSadangi — CHANGES_REQUESTED** [review]

**Review by SiddhantSadangi — APPROVED** [review]

File: src/lightning/pytorch/loggers/neptune.py [hunk 6]
```diff
@@ -43,6 +43,15 @@
 _INTEGRATION_VERSION_KEY = "source_code/integrations/pytorch-lightning"
 
 
+def _get_expected_model_path(dir_path: str) -> str:
```
Comment (Borda) [inline src/lightning/pytorch/loggers/neptune.py:46]: can we add simple doctest to better illustrate what it does

**Review by Borda — APPROVED** [review]

File: src/lightning/pytorch/loggers/neptune.py (hunk 6 above)
Comment (AleksanderWWW) [reply src/lightning/pytorch/loggers/neptune.py:46]: Sure :smile:

**Comment by AleksanderWWW** [comment]
@Borda how shall I proceed with this doctest failure due to the check running on ubuntu, while the line assumes windows os?

File: src/lightning/pytorch/loggers/neptune.py [hunk 7]
```diff
@@ -43,6 +43,26 @@
...
+
+    >>> _sanitize_expected_model_path('my_model/checkpoints//')  # unix-like path separator os
+    'my_model/checkpoints/'
+    >>> _sanitize_expected_model_path('my_model/checkpoints')
+    'my_model/checkpoints/'
+
+    >>> _sanitize_expected_model_path('my_model\\checkpoints')  # windows
+    'my_model\\checkpoints\\'
+
+    """
```
Comment (normandy7) [inline src/lightning/pytorch/loggers/neptune.py:57]:
```suggestion
    """Sanitize the expected model path so that it is correct irrespective of the underlying OS.

    Args:
        dir_path:
            Path of the model.

    Returns:
        Sanitized version of the path.

    Examples:
        >>> _sanitize_expected_model_path('my_model/checkpoints//')  # OS with Unix-like path separator
        'my_model/checkpoints/'
        >>> _sanitize_expected_model_path('my_model/checkpoints')
        'my_model/checkpoints/'
        >>> _sanitize_expected_model_path('my_model\\checkpoints')  # Windows
        'my_model\\checkpoints\\'

    """
```

File: tests/tests_pytorch/loggers/test_neptune.py [hunk 8]
```diff
@@ -23,6 +24,11 @@
 from lightning.pytorch import Trainer
 from lightning.pytorch.demos.boring_classes import BoringModel
 from lightning.pytorch.loggers import NeptuneLogger
+from lightning.pytorch.loggers.neptune import _sanitize_expected_model_path
+
+IS_WINDOWS = sys.platform == "win32"
+skip_if_on_windows = pytest.mark.skipif(IS_WINDOWS, reason="Those tests are specific to non-windows systems")
+skip_if_not_windows = pytest.mark.skipif(not IS_WINDOWS, reason="Those tests are specific to windows os")
```
Comment (normandy7) [inline tests/tests_pytorch/loggers/test_neptune.py:31]:
```suggestion
skip_if_on_windows = pytest.mark.skipif(IS_WINDOWS, reason="These tests are specific to non-Windows systems")
skip_if_not_windows = pytest.mark.skipif(not IS_WINDOWS, reason="These tests are specific to Windows OS")
```

**Review by normandy7 — COMMENTED** [review]
Some styling proposals.

File: src/lightning/pytorch/loggers/neptune.py [hunk 9]
```diff
@@ -551,7 +578,7 @@ def after_save_checkpoint(self, checkpoint_callback: Checkpoint) -> None:
     def _get_full_model_name(model_path: str, checkpoint_callback: Checkpoint) -> str:
         """Returns model name which is string `model_path` appended to `checkpoint_callback.dirpath`."""
         if hasattr(checkpoint_callback, "dirpath"):
-            expected_model_path = f"{checkpoint_callback.dirpath}{os.path.sep}"
+            expected_model_path = _sanitize_expected_model_path(checkpoint_callback.dirpath)
```
Comment (carmocca) [inline src/lightning/pytorch/loggers/neptune.py:581]: Could this simply use https://docs.python.org/3/library/os.path.html#os.path.normpath?

File: src/lightning/pytorch/loggers/neptune.py [hunk 10]
```diff
@@ -551,7 +551,7 @@ def after_save_checkpoint(self, checkpoint_callback: Checkpoint) -> None:
     def _get_full_model_name(model_path: str, checkpoint_callback: Checkpoint) -> str:
         """Returns model name which is string `model_path` appended to `checkpoint_callback.dirpath`."""
         if hasattr(checkpoint_callback, "dirpath"):
-            expected_model_path = f"{checkpoint_callback.dirpath}{os.path.sep}"
+            expected_model_path = os.path.normpath(checkpoint_callback.dirpath) + os.path.sep
```
Comment (SiddhantSadangi) [inline src/lightning/pytorch/loggers/neptune.py:554]: This doesn't seem to work on Windows (traceback: https://github.com/neptune-ai/examples/actions/runs/6691027428/job/18177473941#step:5:419)


```suggestion
            expected_model_path = os.path.normpath(checkpoint_callback.dirpath).replace("/", os.path.sep) + os.path.sep
```

**Review by SiddhantSadangi — CHANGES_REQUESTED** [review]

File: src/lightning/pytorch/loggers/neptune.py (hunk 10 above)
Comment (carmocca) [reply src/lightning/pytorch/loggers/neptune.py:554]: Why does it need to suffix a `/` character anyways? If `os.path.join` is used after, then the separator is added automatically

Comment (carmocca) [reply src/lightning/pytorch/loggers/neptune.py:554]: It is possible that `model_path` needs to be normpath-ed too for that check to pass

**Review by SiddhantSadangi — APPROVED** [review]

File: src/lightning/pytorch/loggers/neptune.py [hunk 11]
```diff
@@ -551,11 +551,12 @@ def after_save_checkpoint(self, checkpoint_callback: Checkpoint) -> None:
     def _get_full_model_name(model_path: str, checkpoint_callback: Checkpoint) -> str:
         """Returns model name which is string `model_path` appended to `checkpoint_callback.dirpath`."""
         if hasattr(checkpoint_callback, "dirpath"):
-            expected_model_path = f"{checkpoint_callback.dirpath}{os.path.sep}"
+            model_path = os.path.normpath(model_path)
+            expected_model_path = f"{os.path.normpath(checkpoint_callback.dirpath)}"
```
Comment (carmocca) [inline src/lightning/pytorch/loggers/neptune.py:555]:
```suggestion
            expected_model_path = os.path.normpath(checkpoint_callback.dirpath)
```

**Review by carmocca — APPROVED** [review]

File: src/lightning/pytorch/loggers/neptune.py (hunk 11 above)
Comment (AleksanderWWW) [reply src/lightning/pytorch/loggers/neptune.py:555]: Ah sorry, brain fog :D

**Comment by awaelchli** [comment]
@AleksanderWWW The tests need to be updated to reflect the change. Here we have two failures:

```
FAILED loggers\test_neptune.py::test_after_save_checkpoint - AssertionError: __getitem__('training/model/checkpoints/model2/with/slashes') call not found
FAILED loggers\test_neptune.py::test_get_full_model_name - AssertionError: assert 'key\\in\\parts' == 'key/in/parts'
  - key/in/parts
  ?    ^  ^
  + key\in\parts
  ?    ^  ^
= 2 failed, 2682 passed, 592 skipped, 6 xfailed, 1 xpassed, 3990 warnings in 652.40s (0:10:52) =
```

Please let me know if you need help with it.

**Comment by AleksanderWWW** [comment]
> @AleksanderWWW The tests need to be updated to reflect the change. Here we have two failures:
> 
> ```
> FAILED loggers\test_neptune.py::test_after_save_checkpoint - AssertionError: __getitem__('training/model/checkpoints/model2/with/slashes') call not found
> FAILED loggers\test_neptune.py::test_get_full_model_name - AssertionError: assert 'key\\in\\parts' == 'key/in/parts'
>   - key/in/parts
>   ?    ^  ^
>   + key\in\parts
>   ?    ^  ^
> = 2 failed, 2682 passed, 592 skipped, 6 xfailed, 1 xpassed, 3990 warnings in 652.40s (0:10:52) =
> ```
> 
> Please let me know if you need help with it.

Hey,
Thank you, on it. Sorry for lack of responsiveness - busy time + some time off on my part. Will try to address it as soon as possible!

**Comment by AleksanderWWW** [comment]
@awaelchli Fixed :) sorry it took so long

**Review by awaelchli — APPROVED** [review]
I added the test case
LGTM

**Comment by awaelchli** [comment]
FYI, there is a related issue here #15705 with other edge cases for this function.
